### PR TITLE
fix both reproduceable failures

### LINKED ISSUE #4588: Conformance failure: Ran out of tries on `suchThatWithTryT`
#4585 fixed `Ran out of tries suchThatT` on `RATIFY`, now we are getting something similar for both `GOV` and `RATIFY` rules:
```
           Ran out of tries (100) on suchThatWithTryT
           
           CallStack (from HasCallStack):
             error, called at src/Constrained/GenT.hs:116:3 in constrained-generators-0.2.0.0-inplace:Constrained.GenT
         

  To rerun use: --match "/Conway/Generators/\"GOV\" input generation time/GovSignal (ConwayEra StandardCrypto) generates within 60000000μs/" --seed 1368988139
```

```
       Ran out of tries (126) on suchThatWithTryT
       
       CallStack (from HasCallStack):
         error, called at src/Constrained/GenT.hs:116:3 in constrained-generators-0.2.0.0-B8kCXVsKW4TJHvOEqTimE3:Constrained.GenT
       (after 23 tests)

  To rerun use: --match "/Conway/MiniTrace/50 MiniTrace tests with trace length of 50/RATIFY/" --seed 882748003
```

### PR DESCRIPTION
**Pull request by TimSheard**
Resolves #4588
Fixed two different problems that result in suchThat failures.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 1]
```diff
@@ -983,13 +985,45 @@ class (HasSpec fn a, Show (Hint a)) => HasGenHint fn a where
...
+conformsToSpecM a spec@(SuspendedSpec v ps) = do
+  ans <- checkPred (singletonEnv v a) ps
+  if ans
+    then pure ()
+    else
+      fatalError
+        ( NE.fromList
+            ["conformsToSpecM SuspendedSpec case", "  " ++ show a, "  (" ++ show spec ++ ")", "fails", ""]
+        )
+conformsToSpecM _ (ErrorSpec es) = fatalError (NE.fromList ("conformsToSpecM ErrorSpec case" : NE.toList es))
```
Comment (lehins) [inline libs/constrained-generators/src/Constrained/Base.hs]: There is a cons operator for NonEmpty:
```suggestion
conformsToSpecM _ (ErrorSpec es) = fatalError ("conformsToSpecM ErrorSpec case" NE.<| es)
```

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 2]
```diff
@@ -3388,32 +3422,31 @@ instance (Ord a, HasSpec fn a) => HasSpec fn (Set a) where
               , "  " ++ show e
               ]
           )
+  -- Special case when elemS is a MemberSpec.
+  -- Just union 'must' with enough elements of 'xs' to meet  'szSpec'
+  genFromTypeSpec (SetSpec must elemS@(MemberSpec xs) szSpec) = do
+    let szSpec' = (szSpec <> geqSpec @fn (sizeOf must) <> maxSpec (cardinality elemS))
```
Comment (lehins) [inline libs/constrained-generators/src/Constrained/Base.hs]:
```suggestion
    let szSpec' = szSpec <> geqSpec @fn (sizeOf must) <> maxSpec (cardinality elemS)
```

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 3]
```diff
@@ -3388,32 +3422,31 @@ instance (Ord a, HasSpec fn a) => HasSpec fn (Set a) where
...
+    choices <- pureGen $ shuffle (xs \\ (Set.toList must))
```
Comment (lehins) [inline libs/constrained-generators/src/Constrained/Base.hs]:
```suggestion
    choices <- pureGen $ shuffle (xs \\ Set.toList must)
```

File: libs/constrained-generators/src/Constrained/GenT.hs [hunk 4]
```diff
@@ -29,7 +29,7 @@
 
 module Constrained.GenT where
 
-import Control.Applicative
+-- import Control.Applicative
```
Comment (lehins) [inline libs/constrained-generators/src/Constrained/GenT.hs]:
```suggestion
```

**Review by lehins — CHANGES_REQUESTED** [review]
Do they have an `Ord` instance? If they do, I'd highly recommend using `nubOrd` instead. We have an implementation of it here, so you can just import it:

https://github.com/IntersectMBO/cardano-ledger/blob/c55f7ca2848d78c029bf182777854af5b71e9562/eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs#L761


The reason for this is `nub` is O(n^2) complexity and is ridiculously slow. We don't want our generators to be even slower then they already are.

*This suggestion goes for all uses of `nub` in this PR*

**Comment by TimSheard** [comment]
So some thoughts about the use of nub.
One of the failures was caused by duplicates in 'xs' in (MemberSpec xs). The nub fixed it.
Further investigation showed, there were lots of duplicates in MemberSpec.
Using nub actually made the code run faster, because the majority of time is spent generating random instances, and the duplicates made this take longer. The time spent simplifying Specifications (where we use the nub) is a small fraction of the total time. The most common MemberSpec has one or two elements.
Almost all  large MemberSpecs arise from big Sets, but in those cases we construct the MemberSpec as follows (MemberSpec (Set.toList set)) where we don't need to use nub. 

In every case where we used nub (I checked this carefully) where we had an Ord instance, I replaced it with nubOrd. In all other contexts, all we have is a (HasSpec fn a) this implies (Eq a) but not (Ord a).  So there is nothing that can be done.

**Review by lehins — APPROVED** [review]
Thank you!!!!

**Pull request merged**